### PR TITLE
[SandboxIR] Added new StoreInst::create() functions with isVolatile arg

### PR DESCRIPTION
**Pull request by wizardengineer**
Closed PR: https://github.com/llvm/llvm-project/pull/100957

As stated this patch just implements the volatile-memory data for StoreInst

### CODE REVIEW

**Comment by medievalghoul** [comment]
fixing the formatting

**Comment by medievalghoul** [comment]
fixing unit test error on CI/CD.

File: llvm/include/llvm/SandboxIR/SandboxIR.h [hunk 1]
```diff
@@ -804,14 +804,23 @@ class StoreInst final : public Instruction {
   }
 
 public:
+  /// Return true if this is a load from a volatile memory location.
+  bool isVolatile() const { return cast<llvm::LoadInst>(Val)->isVolatile(); }
```
Comment (vporpo) [inline llvm/include/llvm/SandboxIR/SandboxIR.h:808]: The build fails because this should be: `cast<llvm::StoreInst>(Val)`, not `llvm::LoadInst`.

Comment (medievalghoul) [reply llvm/include/llvm/SandboxIR/SandboxIR.h]: Yeah I see the issue, this didn't get committed properly. I fixed it.

**Comment by medievalghoul** [comment]
@vporpo after the bots pass, is it okay for you to merge this?

File: llvm/include/llvm/SandboxIR/SandboxIR.h [hunk 2]
```diff
@@ -804,14 +804,23 @@ class StoreInst final : public Instruction {
   }
 
 public:
+  /// Return true if this is a load from a volatile memory location.
```
Comment (vporpo) [inline llvm/include/llvm/SandboxIR/SandboxIR.h:807]: "... if this is a store to ..."

File: llvm/unittests/SandboxIR/SandboxIRTest.cpp [hunk 3]
```diff
@@ -810,10 +815,45 @@ define void @foo(i8 %val, ptr %ptr) {
   sandboxir::StoreInst *NewSt =
       sandboxir::StoreInst::create(Val, Ptr, Align(8),
                                    /*InsertBefore=*/Ret, Ctx);
+  // Check if create was volatile
```
Comment (vporpo) [inline llvm/unittests/SandboxIR/SandboxIRTest.cpp:818]: To keep things consistent we should either remove this comment, or add comments to the rest of the checks. Let's remove it for now. We can add the comments later in a separate PR if needed.

File: llvm/unittests/SandboxIR/SandboxIRTest.cpp [hunk 4]
```diff
@@ -810,10 +815,45 @@ define void @foo(i8 %val, ptr %ptr) {
...
   EXPECT_EQ(NewSt->getPointerOperand(), Ptr);
   EXPECT_EQ(NewSt->getAlign(), 8);
+  // Check create(InsertBefore, IsVolatile=true)
+  sandboxir::StoreInst *NewVSt =
+      sandboxir::StoreInst::create(Val, Ptr, Align(8),
+                                   /*InsertBefore=*/Ret,
+                                   /*IsVolatile=*/true, Ctx);
+  // Check if create was volatile
+  EXPECT_TRUE(NewVSt->isVolatile());
+
```
Comment (vporpo) [inline llvm/unittests/SandboxIR/SandboxIRTest.cpp:831]: I think we should add the same checks as above:
```
  EXPECT_EQ(NewVSt->getType(), St->getType());
  EXPECT_EQ(NewVSt->getValueOperand(), Val);
  EXPECT_EQ(NewVSt->getPointerOperand(), Ptr);
  EXPECT_EQ(NewVSt->getAlign(), 8);
  ```
And also let's also check the instruction position (you can also add this check above too).
```
EXPECT_EQ(NewVSt->getNextNode(), Ret);
```

File: llvm/unittests/SandboxIR/SandboxIRTest.cpp (hunk 3 above)
Comment (medievalghoul) [reply llvm/unittests/SandboxIR/SandboxIRTest.cpp:818]: I'll remove all instance of it.

Comment (medievalghoul) [reply llvm/unittests/SandboxIR/SandboxIRTest.cpp]: fixed.

File: llvm/unittests/SandboxIR/SandboxIRTest.cpp (hunk 4 above)
Comment (medievalghoul) [reply llvm/unittests/SandboxIR/SandboxIRTest.cpp]: I made sure to add those.

File: llvm/unittests/SandboxIR/SandboxIRTest.cpp [hunk 5]
```diff
@@ -785,8 +785,9 @@ define void @foo(ptr %arg0, ptr %arg1) {
 
 TEST_F(SandboxIRTest, StoreInst) {
   parseIR(C, R"IR(
-define void @foo(i8 %val, ptr %ptr) {
+define void @foo(i8 %val, ptr %ptr, ptr %vptr) {
```
Comment (vporpo) [inline llvm/unittests/SandboxIR/SandboxIRTest.cpp:788]: No need to use a seaprate pointer, just use `%ptr`.

File: llvm/unittests/SandboxIR/SandboxIRTest.cpp [hunk 6]
```diff
@@ -798,9 +799,13 @@ define void @foo(i8 %val, ptr %ptr) {
   auto *BB = &*F->begin();
   auto It = BB->begin();
   auto *St = cast<sandboxir::StoreInst>(&*It++);
+  auto *VSt = cast<sandboxir::StoreInst>(&*It++);
   auto *Ret = cast<sandboxir::ReturnInst>(&*It++);
 
   // Check that the StoreInst has been created correctly.
+  // Check isVolatile()
```
Comment (vporpo) [inline llvm/unittests/SandboxIR/SandboxIRTest.cpp:806]: Remove the `// Check isVolatile()` comment.

File: llvm/unittests/SandboxIR/SandboxIRTest.cpp [hunk 7]
```diff
@@ -810,10 +815,45 @@ define void @foo(i8 %val, ptr %ptr) {
   sandboxir::StoreInst *NewSt =
       sandboxir::StoreInst::create(Val, Ptr, Align(8),
                                    /*InsertBefore=*/Ret, Ctx);
+  EXPECT_FALSE(NewSt->isVolatile());
   EXPECT_EQ(NewSt->getType(), St->getType());
   EXPECT_EQ(NewSt->getValueOperand(), Val);
   EXPECT_EQ(NewSt->getPointerOperand(), Ptr);
   EXPECT_EQ(NewSt->getAlign(), 8);
```
Comment (vporpo) [inline llvm/unittests/SandboxIR/SandboxIRTest.cpp:822]: Add a check: `EXPECT_EQ(NewSt->getNextNode(), Ret)`

File: llvm/include/llvm/SandboxIR/SandboxIR.h [hunk 8]
```diff
@@ -804,14 +804,23 @@ class StoreInst final : public Instruction {
   }
 
 public:
+  /// Return true if this is a store from a volatile memory location.
+  bool isVolatile() const { return cast<llvm::StoreInst>(Val)->isVolatile(); }
+
```
Comment (vporpo) [inline llvm/include/llvm/SandboxIR/SandboxIR.h:809]: Remove new line.

**Review by vporpo — COMMENTED** [review]
A few minor changes. Looks good.

File: llvm/unittests/SandboxIR/SandboxIRTest.cpp (hunk 6 above)
Comment (medievalghoul) [reply llvm/unittests/SandboxIR/SandboxIRTest.cpp:806]: in a separate NFC PR, should i remove these comments off LoadInst tests too?

File: llvm/unittests/SandboxIR/SandboxIRTest.cpp (hunk 7 above)
Comment (medievalghoul) [reply llvm/unittests/SandboxIR/SandboxIRTest.cpp:821]: fixed.

File: llvm/unittests/SandboxIR/SandboxIRTest.cpp (hunk 5 above)
Comment (medievalghoul) [reply llvm/unittests/SandboxIR/SandboxIRTest.cpp]: fixed.

File: llvm/include/llvm/SandboxIR/SandboxIR.h (hunk 8 above)
Comment (medievalghoul) [reply llvm/include/llvm/SandboxIR/SandboxIR.h]: fixed.

File: llvm/unittests/SandboxIR/SandboxIRTest.cpp (hunk 6 above)
Comment (medievalghoul) [reply llvm/unittests/SandboxIR/SandboxIRTest.cpp]: Will do, thanks for the suggestion.

**Review by vporpo — APPROVED** [review]
Looks good, thanks for the patch. I will merge it as soon as the bots pass.

**Comment by vporpo** [comment]
Adding `setVerbose()` is a bit trickier because it requires you to track the change with the Tracker. You would need to create a class in `Tracker.h` like `class SetVerbose : public IRChangeBase` containing an `Instruction *I` and a `bool WasVerbose` to track the original state of the Verbose flag.
You can check how this is done in a couple of places, like for example: `CallBrInst::setDefaultDest()`. Take a look at `TrackerTest.cpp` too.

**Comment by medievalghoul** [comment]
I noticed I can't directly request reviewers for SandboxIR contributions. Would it be alright if I continue mentioning you and other relevant reviewers for my patches?

**Comment by vporpo** [comment]
> I noticed I can't directly request reviewers for SandboxIR contributions. 
Hmm what do you mean? Doesn't clicking on "Reviewers" and adding my or other people's name work?

**Comment by medievalghoul** [comment]
> Hmm what do you mean? Doesn't clicking on "Reviewers" and adding my or other people's name work?
It doesn't seems as if I can. I don't think I have the permissions for that.

**Comment by vporpo** [comment]
Don't worry about it, just mention me and I will review the patch.

**Comment by medievalghoul** [comment]
> Adding `setVerbose()` is a bit trickier because it requires you to track the change with the Tracker. You would need to create a class in `Tracker.h` like `class SetVerbose : public IRChangeBase` containing an `Instruction *I` and a `bool WasVerbose` to track the original state of the Verbose flag.
> You can check how this is done in a couple of places, like for example: `CallBrInst::setDefaultDest()`. Take a look at `TrackerTest.cpp` too.

@vporpo Just to be clear, so I'm implementing the setVerbose() method for the StoreInst or is it something completely separate?

**Comment by vporpo** [comment]
@medievalghoul sorry I misspelled: I meant `setVolatile(bool)`

**Comment by medievalghoul** [comment]
No worries, thanks for letting me know.

**Pull request merged**